core/src/blocktree.rs

```diff
@@ -443,22 +448,32 @@ impl Blocktree {
             }
         });
 
-        let recovered_data = Self::try_shred_recovery(
-            &db,
-            &erasure_metas,
-            &index_working_set,
-            &mut just_inserted_data_shreds,
-            &mut just_inserted_coding_shreds,
-        );
-
-        recovered_data.into_iter().for_each(|shred| {
-            self.insert_recovered_data_shred(
-                &shred,
-                &mut index_working_set,
-                &mut slot_meta_working_set,
-                &mut write_batch,
+        if let Some(leader_schedule_cache) = leader_schedule {
+            let mut recovered_data = Self::try_shred_recovery(
+                &db,
+                &erasure_metas,
+                &index_working_set,
+                &mut just_inserted_data_shreds,
+                &mut just_inserted_coding_shreds,
             );
-        });
+
+            recovered_data.retain(|shred| {
+                if let Some(leader) = leader_schedule_cache.slot_leader_at(shred.slot(), None) {
+                    shred.verify(&leader)
```

Comment:
maybe we should make a new `should_insert_recovered_data` function which calls `should_insert_data_shred` (have to double check if all those checks are relevant/need to be updated) and does the pubkey verification.

Reply:
Done..